compiler/rustc_mir_dataflow/src/value_analysis.rs

```diff
@@ -0,0 +1,677 @@
+//!
+//! The default methods of [`ValueAnalysis`] (prefixed with `super_` instead of `handle_`)
+//! provide some behavior that should be valid for all abstract domains that are based only on the
+//! value stored in a certain place. On top of these default rules, an implementation should
+//! override some of the `handle_` methods. For an example, see `ConstAnalysis`.
+//!
+//! An implementation must also provide a [`Map`]. Before the anaylsis begins, all places that
+//! should be tracked during the analysis must be registered. The set of tracked places cannot be
+//! changed during the analysis.
+
+use std::fmt::{Debug, Formatter};
+
+use rustc_data_structures::fx::FxHashMap;
+use rustc_index::vec::IndexVec;
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use rustc_target::abi::VariantIdx;
+
+use crate::{
+    fmt::DebugWithContext, lattice::FlatSet, Analysis, AnalysisDomain, CallReturnPlaces,
+    JoinSemiLattice, SwitchIntEdgeEffects,
+};
+
+pub trait ValueAnalysis<'tcx> {
+    /// For each place of interest, the analysis tracks a value of the given type.
+    type Value: Clone + JoinSemiLattice + HasBottom + HasTop;
+
+    const NAME: &'static str;
+
+    fn map(&self) -> &Map;
+
+    fn handle_statement(&self, statement: &Statement<'tcx>, state: &mut State<Self::Value>) {
+        self.super_statement(statement, state)
+    }
+
+    fn super_statement(&self, statement: &Statement<'tcx>, state: &mut State<Self::Value>) {
+        match &statement.kind {
+            StatementKind::Assign(box (place, rvalue)) => {
+                self.handle_assign(*place, rvalue, state);
+            }
+            StatementKind::SetDiscriminant { .. } => {
+                // Could treat this as writing a constant to a pseudo-place.
+            }
+            StatementKind::CopyNonOverlapping(..) => {
+                // FIXME: What to do here?
+            }
+            StatementKind::StorageLive(..)
+            | StatementKind::StorageDead(..)
+            | StatementKind::Deinit(_) => {
+                // Could perhaps use these.
+            }
+            StatementKind::Nop
+            | StatementKind::Retag(..)
+            | StatementKind::FakeRead(..)
+            | StatementKind::Coverage(..)
+            | StatementKind::AscribeUserType(..) => (),
+        }
+    }
+
+    fn handle_assign(
+        &self,
+        target: Place<'tcx>,
+        rvalue: &Rvalue<'tcx>,
+        state: &mut State<Self::Value>,
+    ) {
+        self.super_assign(target, rvalue, state)
+    }
+
+    fn super_assign(
+        &self,
+        target: Place<'tcx>,
+        rvalue: &Rvalue<'tcx>,
+        state: &mut State<Self::Value>,
+    ) {
+        match rvalue {
+            Rvalue::Ref(_, BorrowKind::Shared, place) => {
+                let target_deref = self
+                    .map()
+                    .find(target.as_ref())
+                    .and_then(|target| self.map().apply_elem(target, ProjElem::Deref));
```

Comment:
why does the target place get dereferenced? that seems odd. `Ref` takes a reference to `place` and assigns that address to `target`

Reply:
Oh... I understand now. You are registering that `*target` refers to `place`. Later uses of `*target` can read from `place` directly.

I think we'll need some high level docs of the place handling scheme in this file

Reply:
I added some documentation regarding the tracking of references. Let me know if we need more (although I will probably add more documentation soon anyways).